### PR TITLE
MouseRegion enter/exit event can be triggered with button pressed

### LINKED ISSUE #69458: MouseRegion does not behave properly while dragging a Draggable over a MouseRegion
## Steps to Reproduce

I have only tested this in Flutter Web:
1. Create a Draggable widget and a MouseRegion that tracks onExit and onEnter.
2. Drag the draggable over the mouse region.

Here's the code to reproduce the error:
```
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: Scaffold(
          body: Center(child: Container(width: 500, child: HoverBug()))),
    );
  }
}

class HoverBug extends StatefulWidget {
  @override
  _HoverBugState createState() => _HoverBugState();
}

class _HoverBugState extends State<HoverBug> {
  bool _mouseOver = false;

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        Text('Mouse is over mouse region (green): ${_mouseOver.toString()} '),
        Row(children: [
          MyDraggable(),
          Container(
            width: 200,
          ),
          MouseRegion(
            onEnter: (event) => setState(() => _mouseOver = true),
            onExit: (event) => setState(() => _mouseOver = false),
            child: Container(
              width: 100,
              height: 100,
              color: Colors.green,
            ),
          )
        ])
      ],
    );
  }
}

class MyDraggable extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Draggable(
      feedback: this,
      child: Container(
        width: 100,
        height: 100,
        color: Colors.blue,
      ),
    );
  }
}

```

**Expected results:** 
Expected onEnter to be called once and onExit to not be called until the mouse exits the MouseRegion.

**Actual results:** 
Instead onExit is called continuously while the mouse is moving over the region and onEnter is called every time the mouse stops while being over the MouseRegion.

<details>
  <summary>Logs</summary>

```
flutter run --verbose
[ +170 ms] executing: [/Users/jonathanaird/developer/flutter sdk/flutter/] git -c
log.showSignature=false log -n 1 --pretty=format:%H
[ +100 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[        ] 198df796aa80073ef22bdf249e614e2ff33c6895
[   +1 ms] executing: [/Users/jonathanaird/developer/flutter sdk/flutter/] git tag --points-at HEAD
[  +95 ms] Exit code 0 from: git tag --points-at HEAD
[        ] 1.23.0-18.1.pre
[  +18 ms] executing: [/Users/jonathanaird/developer/flutter sdk/flutter/] git rev-parse --abbrev-ref
--symbolic @{u}
[  +18 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/beta
[        ] executing: [/Users/jonathanaird/developer/flutter sdk/flutter/] git ls-remote --get-url
origin
[  +17 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[  +75 ms] executing: [/Users/jonathanaird/developer/flutter sdk/flutter/] git rev-parse --abbrev-ref
HEAD
[  +24 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] beta
[  +16 ms] executing: sw_vers -productName
[  +53 ms] Exit code 0 from: sw_vers -productName
[        ] Mac OS X
[        ] executing: sw_vers -productVersion
[  +32 ms] Exit code 0 from: sw_vers -productVersion
[        ] 10.15.7
[        ] executing: sw_vers -buildVersion
[  +25 ms] Exit code 0 from: sw_vers -buildVersion
[        ] 19H2
[        ] executing: uname -m
[  +13 ms] Exit code 0 from: uname -m
[   +1 ms] x86_64
[ +152 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +6 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[ +125 ms] executing: /Users/jonathanaird/Library/Android/sdk/platform-tools/adb devices -l
[   +9 ms] executing: /usr/bin/xcode-select --print-path
[  +14 ms] Exit code 0 from: /usr/bin/xcode-select --print-path
[        ] /Library/Developer/CommandLineTools
[   +2 ms] executing: /usr/bin/xcodebuild -version
[   +4 ms] Exit code 1 from: /usr/bin/xcodebuild -version
[        ] xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory
'/Library/Developer/CommandLineTools' is a command line tools instance
[        ] Xcode not found. Run 'flutter doctor' for more information.
[   +1 ms] executing: /usr/bin/xcodebuild -version
[  +20 ms] Exit code 1 from: /usr/bin/xcodebuild -version
[        ] xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory
'/Library/Developer/CommandLineTools' is a command line tools instance
[  +70 ms] List of devices attached
[   +6 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[   +3 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[  +45 ms] executing: /usr/bin/xcodebuild -version
[   +4 ms] Exit code 1 from: /usr/bin/xcodebuild -version
[        ] xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory
'/Library/Developer/CommandLineTools' is a command line tools instance
[  +72 ms] Multiple devices found:
[ +226 ms] Web Server (web) • web-server • web-javascript • Flutter Tools
[   +1 ms] Chrome (web)     • chrome     • web-javascript • Google Chrome 86.0.4240.111
[   +2 ms] [0]: Web Server (web-server)
[        ] [1]: Chrome (chrome)
[   +3 ms] Please choose one (To quit, press "q/Q")
[        ] : 
[+12658 ms] C
[        ] Please choose one (To quit, press "q/Q")
[        ] : 
[ +251 ms] h
[        ] Please choose one (To quit, press "q/Q")
[        ] : 
[ +155 ms] r
[        ] Please choose one (To quit, press "q/Q")
[        ] : 
[ +116 ms] o
[        ] Please choose one (To quit, press "q/Q")
[        ] : 
[+7094 ms] 1
[ +227 ms] Generating
/Users/jonathanaird/Developer/temp/bugs/mouseregion_draggable_bug/android/app/src/main/java/io/flutte
r/plugins/GeneratedPluginRegistrant.java
[ +110 ms] Launching lib/main.dart on Chrome in debug mode...
[ +163 ms] Updating assets
[ +133 ms] Syncing files to device Chrome...
[  +75 ms] Generating
/Users/jonathanaird/Developer/temp/bugs/mouseregion_draggable_bug/android/app/src/main/java/io/flutte
r/plugins/GeneratedPluginRegistrant.java
[  +67 ms] <- reset
[  +21 ms] /Users/jonathanaird/developer/flutter sdk/flutter/bin/cache/dart-sdk/bin/dart
--disable-dart-dev /Users/jonathanaird/developer/flutter
sdk/flutter/bin/cache/artifacts/engine/darwin-x64/frontend_server.dart.snapshot --sdk-root
/Users/jonathanaird/developer/flutter sdk/flutter/bin/cache/flutter_web_sdk/ --incremental
--target=dartdevc --debugger-module-names --experimental-emit-debug-metadata
-Ddart.developer.causal_async_stacks=true --output-dill
/var/folders/6h/ch5qgxw90r99kmpgcvpvlc8h0000gn/T/flutter_tools.mlsnqA/flutter_tool.4ZxFUe/app.dill
--libraries-spec
file:///Users/jonathanaird/developer/flutter%20sdk/flutter/bin/cache/flutter_web_sdk/libraries.json
--packages .packages -Ddart.vm.profile=false -Ddart.vm.product=false --enable-asserts
--track-widget-creation --filesystem-root
/var/folders/6h/ch5qgxw90r99kmpgcvpvlc8h0000gn/T/flutter_tools.mlsnqA/flutter_tools.joNvLB
--filesystem-scheme org-dartlang-app --initialize-from-dill build/cache.dill.track.dill --platform
file:///Users/jonathanaird/developer/flutter%20sdk/flutter/bin/cache/flutter_web_sdk/kernel/flutter_d
dc_sdk.dill
[  +17 ms] <- compile org-dartlang-app:/web_entrypoint.dart
[+24439 ms] Syncing files to device Chrome... (completed in 24,596ms, longer than expected)
[   +1 ms] Synced 28.1MB.
[        ] <- accept
[        ] Caching compiled dill
[ +367 ms] Using Google Chrome 86.0.4240.111 

[ +839 ms] [CHROME]:
[   +2 ms] [CHROME]:DevTools listening on
ws://127.0.0.1:57229/devtools/browser/58a74f93-34b7-449e-b119-45fdc23fc2fc
[+5879 ms] Debug service listening on ws://127.0.0.1:57251/n2Qr7KcNQHM=

[ +287 ms] Debug service listening on ws://127.0.0.1:57251/n2Qr7KcNQHM=
[   +2 ms] Warning: Flutter's support for web development is not stable yet and hasn't
[        ] been thoroughly tested in production environments.
[        ] For more information see https://flutter.dev/web
[        ] 🔥  To hot restart changes while running, press "r" or "R".
[   +1 ms] For a more detailed help message, press "h". To quit, press "q".
```

```
flutter analyze
Analyzing mouseregion_draggable_bug...                                  
No issues found! (ran in 2.9s)
```

```
[✓] Flutter (Channel beta, 1.23.0-18.1.pre, on Mac OS X 10.15.7 19H2 x86_64, locale en-US)
    • Flutter version 1.23.0-18.1.pre at /Users/jonathanaird/developer/flutter sdk/flutter
    • Framework revision 198df796aa (2 weeks ago), 2020-10-15 12:04:33 -0700
    • Engine revision 1d12d82d9c
    • Dart version 2.11.0 (build 2.11.0-213.1.beta)

 
[!] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/jonathanaird/Library/Android/sdk
    ✗ Flutter requires Android SDK 29 and the Android BuildTools 28.0.3
      To update the Android SDK visit
      https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions.
    • All Android licenses accepted.

[✗] Xcode - develop for iOS and macOS
    ✗ Xcode installation is incomplete; a full installation is necessary for iOS development.
      Download at: https://developer.apple.com/xcode/download/
      Or install Xcode via the App Store.
      Once installed, run:
        sudo xcode-select --switch /Applications/Xcode.app/Contents/Developer
        sudo xcodebuild -runFirstLaunch
    ✗ CocoaPods installed but not working.
        You appear to have CocoaPods installed but it is not working.
        This can happen if the version of Ruby that CocoaPods was installed with is different from
        the one being used to invoke it.
        This can usually be fixed by re-installing CocoaPods. For more info, see
        https://github.com/flutter/flutter/issues/14293.
      To re-install CocoaPods, run:
        sudo gem install cocoapods

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 33.3.1
    • Dart plugin version 182.5215
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)

[✓] VS Code (version 1.50.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (2 available)
    • Web Server (web) • web-server • web-javascript • Flutter Tools
    • Chrome (web)     • chrome     • web-javascript • Google Chrome 86.0.4240.111

! Doctor found issues in 2 categories.
```

</details>

### LINKED ISSUE #80942: MouseRegion rebuilds every time when the pointer does not tap up
If you tap down and hold it (without releasing the pointer), MouseRegion stops behaving properly, and it seems that it also breaks if it builds while the pointer was never released.

Here's a reproducible example in dart pad: https://dartpad.dev/719b5207a759f3f6c39947dd888bda1d?null_safety=true

Steps to reproduce: 
- When the `Click me!` text appears, tap down on it and do NOT tap up on it.
- Move the mouse WITHIN the bounds of the red box that appears, you will notice the EXIT text flickers.

This is because onExit keeps repeatedly firing for every mouse movement, even though the mouse is INSIDE of the red box.

In this gif, I do NOT remove my finger off of the left mouse button at any point after clicking on the `Click me!` text.

Additional steps:
- Tap down inside the red box and do not tap up.
- Move the pointer outside of the box and you will notice the EXIT text flickers again. 

The pointer left the box once, but the MouseRegion.onExit function fires every time the mouse moves outside the box with the pointer staying tapped down.

In this gif, I do NOT remove my finger off of the left mouse button at any point after entering the red box the first time.

This issue with MouseRegion is critical and currently affecting our team's project when handling some very complex free-form interaction on a custom canvas-like area.

### PR DESCRIPTION
**Pull request by xu-baolin**
Fixes #69458 
Fixes #80942

The doc says that event can be triggered with or without buttons pressed:
```dart
  /// Triggered when a mouse pointer has entered this widget.
  ///
  /// This callback is triggered when the pointer, with or without buttons
  /// pressed, has started to be contained by the region of this widget. More
  /// specifically, the callback is triggered by the following cases:
```

### Code Sample
```dart
  testWidgets('onEnter and onExit can be triggered with mouse buttons pressed', (WidgetTester tester) async {
    PointerEnterEvent? enter;
    PointerExitEvent? exit;
    await tester.pumpWidget(Center(
      child: MouseRegion(
        child: Container(
          color: const Color.fromARGB(0xff, 0xff, 0x00, 0x00),
          width: 100.0,
          height: 100.0,
        ),
        onEnter: (PointerEnterEvent details) => enter = details,
        onExit: (PointerExitEvent details) => exit = details,
      ),
    ));

    final TestGesture gesture = await tester.createGesture(kind: PointerDeviceKind.mouse, buttons: kPrimaryMouseButton);
    await gesture.addPointer(location: Offset.zero);
    await gesture.down(Offset.zero); // Press the mouse button.
    addTearDown(gesture.removePointer);
    await tester.pump();
    enter = null;
    exit = null;
    // Trigger the enter event.
    await gesture.moveTo(const Offset(400.0, 300.0));
    expect(enter, isNotNull);
    expect(enter!.position, equals(const Offset(400.0, 300.0)));
    expect(enter!.localPosition, equals(const Offset(50.0, 50.0)));
    expect(exit, isNull);

    // Trigger the exit event.
    await gesture.moveTo(const Offset(1.0, 1.0));
    expect(exit, isNotNull);
    expect(exit!.position, equals(const Offset(1.0, 1.0)));
    expect(exit!.localPosition, equals(const Offset(-349.0, -249.0)));
  });
```

### CODE REVIEW

File: packages/flutter/lib/src/gestures/binding.dart [hunk 1]
```diff
@@ -355,12 +355,19 @@ mixin GestureBinding on BindingBase implements HitTestable, HitTestDispatcher, H
     } else if (event is PointerUpEvent || event is PointerCancelEvent) {
       hitTestResult = _hitTests.remove(event.pointer);
     } else if (event.down) {
-      // Because events that occur with the pointer down (like
-      // [PointerMoveEvent]s) should be dispatched to the same place that their
-      // initial PointerDownEvent was, we want to re-use the path we found when
-      // the pointer went down, rather than do hit detection each time we get
-      // such an event.
-      hitTestResult = _hitTests[event.pointer];
+      if (event.kind == PointerDeviceKind.mouse) {
```
Comment (xu-baolin) [inline packages/flutter/lib/src/gestures/binding.dart:358]: Actually, I want to replace all the logic with this, but I can't fully evaluate the impact, so I limit the changes to mouse operations.

**Comment by dkwingsmt** [comment]
Thanks for the fix, although I don't think this is the best way to fix it. You changed to whom the Move events are dispatched to fix enter/exit. In the current event model, the hover/down/up/move events are independent from enter/exit. The root cause of this problem is how the target annotation for MouseTracker are calculated.

My proposal is to change `RenderingBinding.dispatchEvent` to:
```dart
  @override // from GestureBinding
  void dispatchEvent(PointerEvent event, HitTestResult? hitTestResult) {
    if (hitTestResult != null ||
        event is PointerAddedEvent ||
        event is PointerRemovedEvent) {
      assert(event.position != null);
      _mouseTracker!.updateWithEvent(event, () =>
        (event.down || hitTestResult == null) ? renderView.hitTestMouseTrackers(event.position) : hitTestResult
      );
    }
    super.dispatchEvent(event, hitTestResult);
  }
```

(Feel free to incorporate this change to your PR if you agree :)  )

**Review by dkwingsmt — APPROVED** [review]
LGTM

**Comment by ToyWalrus** [comment]
Just curious, what caused this PR to be reverted?

**Comment by xu-baolin** [comment]
> Just curious, what caused this PR to be reverted?

Still investigating, we will try to reland this ASAP.

**Pull request merged**